lib/Sema/CSApply.cpp

```diff
@@ -5730,13 +5749,35 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
           arg, closureType->getResult(),
           locator.withPathElement(ConstraintLocator::AutoclosureResult));
 
+      if (shouldInjectWrappedValuePlaceholder) {
+        // If the autoclosure is escaping, we want the placeholder to wrap
+        // an autoclosure expression.
+        if (!closureType->isNoEscape()) {
+          auto *placeholder = injectWrappedValuePlaceholder(
+              cs.buildAutoClosureExpr(arg, closureType));
+
+          // We can't pass an opaque closure directly as an autoclosure arg.
```

Comment:
Is there really a requirement after type checking that the argument to an @autoclosure parameter is syntactically an AutoClosureExpr?

Reply:
I think the comment isn't properly explaining why this code is doing all of this wrapping. If the opaque value is a placeholder for an autoclosure (which needs to be the case when `init(wrappedValue:)` takes an escaping autoclosure), we want the affect of autoclosure forwarding here, and the only way to do that is to call the autoclosure when it's passed to the next function (see https://github.com/apple/swift/pull/30537)